### PR TITLE
feat(datasource/dotnet): add dotnet datasource

### PR DESCRIPTION
**Pull request by JamieMagee**
## Changes

Add `dotnet` datasource

## Context

#11342

NOTE: This adds the datasource, but does not hook it up to be used.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by JamieMagee** [comment]
FYI @fgreinacher

File: lib/modules/datasource/dotnet-sdk/index.ts
Comment (rarkins) [inline lib/modules/datasource/dotnet-sdk/index.ts:42]: Make sure this is an array, e.g. using `coerceArray()`

File: lib/modules/datasource/dotnet-sdk/index.ts
Comment (rarkins) [inline lib/modules/datasource/dotnet-sdk/index.ts:73]: Also harden this for non-array responses

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/modules/datasource/dotnet-sdk/index.spec.ts
Comment (viceice) [inline lib/modules/datasource/dotnet-sdk/index.spec.ts:11]: can we inline those and reduce to minimum?

because IO is expensive on CI (especially on windows)

File: lib/modules/datasource/dotnet-sdk/index.ts
Comment (viceice) [inline lib/modules/datasource/dotnet-sdk/index.ts:9]: i think this should support runtime too. so we can use this datasource for dotnet runtime updates later.

the adoption datasource for java is a good sample for jre and jdk

File: lib/modules/datasource/dotnet-sdk/index.ts
Comment (viceice) [reply lib/modules/datasource/dotnet-sdk/index.ts:42]: we should use the new JSON Schema validation feature here

we should also cache the sub release files much longer and only get h them when they changed, can be easily done by checking the main release file.

the old EOL versions will no longer change

File: lib/modules/datasource/dotnet-sdk/index.spec.ts
Comment (JamieMagee) [reply lib/modules/datasource/dotnet-sdk/index.spec.ts:11]: I can definitely reduce them down to the properties we need, but even still they are relatively large. I want to keep them as separate files.

File: lib/modules/datasource/dotnet-sdk/index.ts
Comment (JamieMagee) [reply lib/modules/datasource/dotnet-sdk/index.ts:42]: I didn't realise that had landed! Happy to move over to it.

File: lib/modules/datasource/dotnet-sdk/index.ts
Comment (JamieMagee) [reply lib/modules/datasource/dotnet-sdk/index.ts:73]: Closing as not required with the JSON Schema validation approach.

File: lib/modules/datasource/dotnet-sdk/index.ts
Comment (JamieMagee) [reply lib/modules/datasource/dotnet-sdk/index.ts:9]: I'll rename this to `dotnet` instead of `dotnet-sdk`?

File: lib/modules/datasource/dotnet-sdk/index.spec.ts
Comment (JamieMagee) [reply lib/modules/datasource/dotnet-sdk/index.spec.ts:11]: I am stripping out all unnecessary properties using `jq`. For posterity, here's the (monstrous 😱) `jq` filter I used:

```
. | 
{
    "channel-version",
    "latest-release",
    "latest-release-date",
    "latest-runtime",
    "latest-sdk",
    "support-phase",
    releases: [.releases[] |
    { 
        "release-date",
        "release-version",
        security,
        "release-notes",
        runtime: {
            "version": .runtime.version,
            "version-display": .runtime."version-display"
        },
        sdk: {
            "version": .sdk.version,
            "version-display": .sdk."version-display"
        },
        
    }]
}
```

File: lib/modules/datasource/dotnet-sdk/index.ts
Comment (JamieMagee) [reply lib/modules/datasource/dotnet-sdk/index.ts]: I've moved over to JSON schema validation (runtime type checking), added a cache for the release channel fetching, and set the cache time to 1440 minutes (1 day) for both caches.

File: lib/modules/datasource/dotnet-sdk/index.ts
Comment (JamieMagee) [reply lib/modules/datasource/dotnet-sdk/index.ts]: Renamed to `dotnet` and added support for `dotnet-sdk` and `dotnet-runtime` dependencies.

File: lib/modules/datasource/dotnet/__fixtures__/releases-3.1.json
Comment (viceice) [inline lib/modules/datasource/dotnet/__fixtures__/releases-3.1.json:8]: i think we can reduce these files some more and only leave a couple of versions (eg 3 or 5)

File: lib/modules/datasource/dotnet/index.spec.ts
Comment (viceice) [inline lib/modules/datasource/dotnet/index.spec.ts:18]: why do we need this? should be required when tests are good 😉

File: lib/modules/datasource/dotnet/schema.ts
Comment (viceice) [inline lib/modules/datasource/dotnet/schema.ts:3]: should we name those cont with `Schema` suffix and remove the `Type` suffix from types?

🤔

Comment (JamieMagee) [reply lib/modules/datasource/dotnet/schema.ts:3]: Yeah, can do. I think this is the first datasource to use JSON schemas, so I am still figuring out the best practices.

File: lib/modules/datasource/dotnet/__fixtures__/releases-3.1.json
Comment (JamieMagee) [reply lib/modules/datasource/dotnet/__fixtures__/releases-3.1.json:8]: I cut each fixture down and tried to include:
- First and last `preview` release
- One `rc` release
- First and last GA release

**Comment by JamieMagee** [comment]
@viceice can I get another review please?

**Review by viceice — APPROVED** [review]

**Comment by JamieMagee** [comment]
@rarkins blocked on your review

**Review by rarkins — APPROVED** [review]

**Pull request merged**
